Generalize access logging (rebased)

jrajahalme opened the pull request:
Refactor access logging internals from oxyproxy.go to accesslog/log.go, allowing for sharing most of the code with other proxy types.

Note that the (marked internal) http.Request has been removed from the access log message. This means that it is also removed from the json formatted output.

aanm commented on pkg/proxy/accesslog/log.go:
```diff
@@ -127,3 +204,102 @@ func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
+// egressDestinationInfo returns the destination EndpointInfo for a flow
+// leaving the proxy at egress.
+func (info *EndpointInfo) egressDestinationInfo(ipstr string) {
+	ip := net.ParseIP(ipstr)
```
I've seen this code in PR #1870 @manalibhutiyani or @jrajahalme is this duplicated or related?

tgraf replied:
We can rebase 1870 on top of this, 1870 is currently duplicating the http code for kafka

manalibhutiyani replied:
I have removed the duplication actually in 1870

raybejjani commented on pkg/proxy/accesslog/log.go:
```diff
@@ -88,33 +97,94 @@ func logString(outStr string, retry bool) {
 	// to reopen the logfile and parallel writes could fail because of that
+
+	l.Timestamp = time.Now().UTC().Format(time.RFC3339Nano)
+	l.Info = info
```
Should these two writes be guarded by `logMutex`? This is the only place we use it and from the comment it seems to be a coordination lock more than a resource lock. If so, it might make sense to lock just before `logString` instead.

jrajahalme replied:
These writes were done before taking the lock by the caller before this PR, so this PR just preserves the previous locking logic.

Looking closer I see the mutex is needed to serialize access to the log buffer and file. The caller owns the request, so there is no contention for accessing it. We could actually move the line taking a log a bit later in the Log() function.

jrajahalme replied:
Also noticed that OpenLogfile() and SetMetadata() did not take the lock. Not a problem currently, as they are only called from sync.Once(), but potential problem later, so I cleaned that up as well.

jrajahalme replied:
I made the changes you requested, maybe review again?

raybejjani replied:
A+1 with extra credit :) Sorry for forcing you to track down all the callpaths to check locking :(

raybejjani commented on pkg/proxy/accesslog/log.go:
```diff
@@ -134,3 +204,102 @@ func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
+// leaving the proxy at egress.
+func (info *EndpointInfo) egressDestinationInfo(ipstr string) {
+	ip := net.ParseIP(ipstr)
+	if ip != nil {
```
This could be `if ip == nil { return }` if you wanted to reduce the indent by one.

jrajahalme replied:
OK, thanks!

raybejjani reviewed (changes requested):
Another proxy owner should probably take a look too, but this is +1 overall. One thing about locks, however.

rlenglet commented on pkg/proxy/accesslog/log.go:
```diff
@@ -17,10 +17,19 @@ package accesslog
+	"strconv"
+	"time"
 
+	"github.com/cilium/cilium/common/addressing"
```
Package github.com/cilium/cilium/pkg/proxy/accesslog is a public API. It's used (vendored) by third-party software. Adding more transitive dependencies should be avoided. Please remove those additional dependencies. If that means moving the code depending on this out of github.com/cilium/cilium/pkg/proxy/accesslog, do that.

jrajahalme replied:
OK, was not aware of this. OK to remove the http.Request from the log record, though?

tgraf reviewed (changes requested):
#1824 is overlapping with this, please hold off on merging

jrajahalme committed:
proxy: Remove private records from access log record.

Envoy integration is easier without the http.Record in the access log.

Move HTTP logging code to a new file proxy/logrecord.go so that it can
also be used for Envoy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
